### ISSUE TITLE
rpi-4.8.y dwc_otg build issues

### ISSUE THREAD

**Issue by igoropaniuk** [raspberrypi/linux#1779]
I'm encountering compilation issues in dwc_otg kmod, while trying to build rpi-4.8.y.

```
In file included from drivers/usb/host/dwc_otg/dwc_otg_driver.c:51:0:
drivers/usb/host/dwc_otg/dwc_otg_os_dep.h:79:26: fatal error: asm/mach/map.h: No such file or directory
compilation terminated.
scripts/Makefile.build:289: recipe for target 'drivers/usb/host/dwc_otg/dwc_otg_driver.o' failed
make[5]: *** [drivers/usb/host/dwc_otg/dwc_otg_driver.o] Error 1
scripts/Makefile.build:440: recipe for target 'drivers/usb/host/dwc_otg' failed
make[4]: *** [drivers/usb/host/dwc_otg] Error 2
scripts/Makefile.build:440: recipe for target 'drivers/usb/host' failed
make[3]: *** [drivers/usb/host] Error 2
scripts/Makefile.build:440: recipe for target 'drivers/usb' failed
make[2]: *** [drivers/usb] Error 2
Makefile:970: recipe for target 'drivers' failed
make[1]: *** [drivers] Error 2
make[1]: Leaving directory '/home/xdev/work/linaro/reps/rpi3_repo/linux'
common.mk:148: recipe for target 'linux-common' failed
```

I'm using default https://github.com/raspberrypi/linux/blob/rpi-4.8.y/arch/arm64/configs/bcmrpi3_defconfig config, but instead of `dwc2` USB controller driver driver trying to compile `dwc_otg`.

```
@@ -955,7 +1195,6 @@ CONFIG_USB_MICROTEK=m
 CONFIG_USBIP_CORE=m
 CONFIG_USBIP_VHCI_HCD=m
 CONFIG_USBIP_HOST=m
-CONFIG_USB_DWC2=y
+CONFIG_USB_DWCOTG=y
```
Maybe someone has encountered similar issues?

**Comment by pelwell** [comment]
64-bit kernel builds (which is what bcmrpi3_defconfig will produce) aren't currently supported by Raspberry Pi. I'm happy for other people to chime in with suggestions, but I suspect that DWC2 is selected by bcmrpi3_defconfig precisely because DWCOTG doesn't build.

**Comment by pelwell** [comment]
You also need to be aware that the Linux ARM64 architecture code doesn't include FIQ support, so even if you get it to build and run performance won't be very good.

**Comment by igoropaniuk** [comment]
Hi @pelwell, thanks for your reply.

I thought that this PR https://github.com/raspberrypi/linux/pull/1611/commits/4d526e25dab0447875855349cb0a6701dd8c8aa8 is going to be merged to the next kernel versions also. (frankly speaking, I haven't checked if these patches exist in 4.8, before I started playing with it). I've successfully built and used 4.6 kernel with DWCOTG for ARM64 before.

But anyway, the reason I decided to use DWCOTG instead of DWC2 is that I can't manage to get wired networking working during kernel boot (I need to mount the root FS via NFS). That's all I receive in kernel log (related to dwc2) during boot:
```
[   19.011609] dwc2 3f980000.usb: Configuration mismatch. dr_mode forced to host
[   19.176845] dwc2 3f980000.usb: 256 invalid for host_nperio_tx_fifo_size. Check HW configuration.
[   19.177069] dwc2 3f980000.usb: 512 invalid for host_perio_tx_fifo_size. Check HW configuration.
[   19.277237] dwc2 3f980000.usb: DWC OTG Controller
[   19.277925] dwc2 3f980000.usb: new USB bus registered, assigned bus number 1
[   19.278746] dwc2 3f980000.usb: irq 41, io mem 0x00000000
```
But. it starts working after some period of time (pay attention on timestamp values). To tell the truth, I can't explain such strange behavior:
```
[ 5807.013163] usb usb1-port1: Cannot enable. Maybe the USB cable is bad?
[ 5807.584891] usb 1-1: new high-speed USB device number 3 using dwc2
[ 5808.163958] usb 1-1: New USB device found, idVendor=0424, idProduct=9514
[ 5808.273156] usb 1-1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[ 5808.402019] hub 1-1:1.0: USB hub found
[ 5808.466637] hub 1-1:1.0: 5 ports detected
[ 5809.096923] usb 1-1.1: new high-speed USB device number 4 using dwc2
[ 5809.563778] usb 1-1.1: New USB device found, idVendor=0424, idProduct=ec00
[ 5809.675786] usb 1-1.1: New USB device strings: Mfr=0, Product=0, SerialNumber=0
[ 5809.843495] smsc95xx v1.0.4
[ 5810.172316] smsc95xx 1-1.1:1.0 eth0: register 'smsc95xx' at usb-3f980000.usb-1.1, smsc95xx USB 2.0 Ethernet, b8:27:eb:7c:97:f6
```
What can cause such behavior?

**Comment by igoropaniuk** [comment]
Bootargs I'm using:

```
bootargs=console=ttyS0,115200 root=/dev/nfs rw rootfstype=nfs nfsroot=192.168.1.5:/opt/linaro/nfs,udp,vers=3 ip=dhcp ignore_loglevel dma.dmachans=0x7f35 rootwait 8250.nr_uarts=1 elevator=deadline fsck.repair=yes smsc95xx.macaddr=b8:27:eb:7c:97:f6 bcm2708_fb.fbwidth=1920 bcm2708_fb.fbheight=1080 vc_mem.mem_base=0x3dc00000 vc_mem.mem_size=0x3f000000
```
Just a remark: this bootargs work fine with 4.6 kernel

**Comment by ED6E0F17** [comment]
For reference, I am using dwc2 on a Pi2 with Linux version 4.8.9, and I get:

[    1.649657] dwc2 3f980000.usb: DWC OTG Controller
[    8.703774] smsc95xx 1-1.1:1.0 eth0: link up, 100Mbps, full-duplex, lpa 0x45E1

There may be issues with 64bit, but in principle the dwc2 driver is good. I do not expect to see DWC_OTG ported to 64bit.

**Comment by lategoodbye** [comment]
@igoropaniuk According your dwc2 kernel messages there is a mismatch between your kernel config and dr_mode in devicetree. As i wrote in #1713 there is currently no official support for otg or peripheral mode for bcm2835 in the upstream dwc2 driver. The Synopsys guys still working on that [issue](http://marc.info/?l=linux-usb&m=148223867014874&w=2).

Please try to set dr_mode to host.

**Comment by igoropaniuk** [comment]
Hi @lategoodbye, thanks for your suggestion. 
I've tried specify dr_mode = "host" in usb node in the DT.
It helped to fix this message about configuration mismatch, but unfortunately enumeration issues still exist (facing the same behavior):
```
[   19.011609] dwc2 3f980000.usb: Configuration mismatch. dr_mode forced to host
```

**Comment by lategoodbye** [comment]
These are not enough information to narrow down the issue.
Do you mean with enumeration issue the fact that the USB root is detected very late?
Is the time period (until USB root detected) nearly the same or random?
Did you try the original defconfig from the same directory?
Did you try rpi-4.9?
Does USB work reliable after the USB root has been detected?

Please provide the actual config values of your build for the following settings:
CONFIG_PM
CONFIG_SUSPEND
CONFIG_USB_OTG
CONFIG_USB_DWC2
CONFIG_USB_DWC2_HOST
CONFIG_USB_DWC2_PERIPHERAL
CONFIG_USB_DWC2_DUAL_ROLE
CONFIG_USB_GADGET

**Comment by igoropaniuk** [comment]
Hi @lategoodbye 
> Do you mean with enumeration issue the fact that the USB root is detected very late?

right

> Is the time period (until USB root detected) nearly the same or random?

random

> Did you try the original defconfig from the same directory?

not yet
> Did you try rpi-4.9?

I tried to compile rpi-4.9 for ARM64, but didn't manage to boot it, that's why I've switched to 4.8

> Does USB work reliable after the USB root has been detected?

Frankly speaking, I didn't test USB a lot, but at least I didn't have any issues with networking after NIC was detected.

Config values:
```
CONFIG_PM=y
CONFIG_USB_DWC2=y
CONFIG_USB_DWC2_HOST=y
# CONFIG_USB_DWC2_DEBUG is not set
# CONFIG_USB_DWC2_TRACK_MISSED_SOFS is not set
# CONFIG_USB_OTG is not set
# CONFIG_SUSPEND is not set
# CONFIG_SYNOPSYS_DWC_ETH_QOS is not set
# CONFIG_USB_DWCOTG is not set
# CONFIG_USB_DWC3 is not set
# CONFIG_USB_GADGET is not set
```

**Comment by Electron752** [comment]
I use dwc_otg locally all the time in ARM64.   I just was never able to get it merged into the downstream or upstream code. The trick is to get the irq chip driver to "round robin" dispatch IRQs to take advantage of more cores.  I found that by doing that, I can get usb to work quite well in 64 bit.

Of course, the best path would be to get FIQ to work in arm64.   Or somehow get interrupt priorities and nested interrupts to work in the Linux kernel.  But that's more work.

I'm locally using a 4.9 based downstream kernel with local patches and it works fine.   I might either today or tomorrow see if I can get 4.9 working again on the public github version.

**Comment by Electron752** [comment]
I opened pull request https://github.com/raspberrypi/linux/pull/1794 for my latest attempt at porting dwc_otg to ARM64.

Hopefully it will get accepted.

**Comment by igoropaniuk** [comment]
@Electron752, thanks a lot for your PR 
Going to test it today

**Comment by Electron752** [comment]
@igoropaniuk, I'm very interested to hear about your tests results.   My simple tests can quickly demonstrate that in some cases the dwc_otg driver as implemented in this PR has less limitations then dwc2.

**Comment by Electron752** [comment]
@igoropaniuk: This has been merged into 4.10 and it should build now.  Happy experimenting!

I'm still interested to know how this works for you.

**Comment by igoropaniuk** [comment]
Hi @Electron752 
Sorry for the late answer (I didn't have a possibility to check it before).
The same issue with your patches:

```
...
[   38.717402] usb usb1: New USB device found, idVendor=1d6b, idProduct=0002
[   38.829197] usb usb1: New USB device strings: Mfr=3, Product=2, SerialNumber=1
[   38.947296] usb usb1: Product: DWC OTG Controller
[   39.024253] usb usb1: Manufacturer: Linux 4.10.0-rc6-v8 dwc_otg_hcd
[   39.126432] usb usb1: SerialNumber: 3f980000.usb
[   39.224068] hub 1-0:1.0: USB hub found
[   39.286453] hub 1-0:1.0: 1 port detected
[   39.366190] dwc_otg: FIQ enabled
[   39.419419] dwc_otg: NAK holdoff enabled
[   39.486400] dwc_otg: FIQ split-transaction FSM enabled
...
```

```
root@RPi3:/ uname -a
Linux RPi3 4.10.0-rc6-v8 #1 SMP PREEMPT Sat Feb 4 18:31:41 EET 2017 aarch64 GNU/Linux
root@RPi3:/ lsusb
Bus 001 Device 001: ID 1d6b:0002
```

Best regards,
Igor

**Comment by Electron752** [comment]
I honestly know nothing about using NFS as a root, so I'm can't help much.  Your original issue was that dwc_otg didn't build, and last time I checked it did and is functional for me.

Perhaps you should try a 32 bit build and see if you get the same issue.  That would give us something to compare against.

**Comment by Electron752** [comment]
@igoropaniuk 

I was able to test network boot today, and I'm finding that everything is working on my pi3 running in arm64 mode with dwc_otg as the usb driver.  To test, I used the trick of copying a single bootcode.bin to a fat partition with nothing else instead of blowing the OTP fuses in my pi3.

I did find something that took me a long time to figure out that might be related.  The kernel will default to version 2 of NFS and that doesn't seem to work well with the os on my pc which is debian sid.  So what I had to do is add a command line parameter to force version 3.  If this isn't done, the root isn't mounted properly and I'll get a kernel panic on the pi3.

Here is the cmdline.txt that I used for testing:
coherent_pool=4M console=tty1 root=/dev/nfs ro nfsroot=192.168.1.100:/srv/pi9,vers=3 ip=192.168.1.35

I mount the root filesystem as ro on the command line and remount it as rw in fstab.
